examples/acoustic/gradient_example.py

```diff
@@ -0,0 +1,178 @@
+        autotuning=False, compiler=None, cache_blocking=None):
+    ndim = len(dimensions)
+    origin = tuple([0.] * ndim)
+    f0 = .010
+
+    t0 = 0.0
+    nsrc = 1
+    nrec = 101
+    # True velocity
+    true_vp = np.ones(dimensions) + .5
+    true_vp[:, :, int(dimensions[2] / 2):] = 2.
+
+    # Smooth velocity - we use this as our initial m
+    initial_vp = smooth10(true_vp, dimensions)
+    m0 = initial_vp**-2
+    # Model perturbation
+    dm = true_vp**-2 - initial_vp**-2
+    model = Model(origin, spacing, true_vp.shape, true_vp, nbpml=nbpml)
+    dt = model.critical_dt
+    if time_order == 4:
+        dt *= 1.73
+    dtype = model.dtype
+    h = model.get_spacing()
+    nt = int(1+(tn-t0)/dt)
+
+    # Source geometry
+    time_series = np.zeros((nt, nsrc))
+    time_series[:, 0] = source(np.linspace(t0, tn, nt), f0)
+
+    location = np.zeros((nsrc, 3))
+    location[0, 0] = origin[0] + dimensions[0] * spacing[0] * 0.5
+    location[0, 1] = origin[1] + dimensions[1] * spacing[1] * 0.5
+    location[0, 2] = origin[1] + 2 * spacing[2]
+
+    # Receiver geometry
+    receiver_coords = np.zeros((nrec, 3))
+    receiver_coords[:, 0] = np.linspace(0, origin[0] + dimensions[0] * spacing[0],
+                                        num=101)
+    receiver_coords[:, 1] = origin[1] + dimensions[1] * spacing[1] * 0.5
+    receiver_coords[:, 2] = location[0, 2]
+
+    # Create source symbol
+    p_src = Dimension('p_src', size=nsrc)
+    src = SourceLike(name="src", dimensions=[time, p_src], npoint=nsrc, nt=nt,
+                     dt=dt, h=h, ndim=ndim, nbpml=nbpml, dtype=dtype,
+                     coordinates=location)
+    src.data[:] = time_series
+
+    # Create receiver symbol
+    p_rec = Dimension('p_rec', size=nrec)
+
+    # Receiver for true model
+    rec_t = SourceLike(name="rec_t", dimensions=[time, p_rec], npoint=nrec, nt=nt, dt=dt,
+                       h=h, ndim=ndim, nbpml=nbpml, dtype=dtype,
+                       coordinates=receiver_coords)
+    # Receiver for smoothed model
+    rec_s = SourceLike(name="rec_s", dimensions=[time, p_rec], npoint=nrec, nt=nt, dt=dt,
+                       h=h, ndim=ndim, nbpml=nbpml, dtype=dtype,
+                       coordinates=receiver_coords)
+
+    # Create the forward wavefield to use (only 3 timesteps)
+    # Once checkpointing is in, this will be the only wavefield we need
+    u_nosave = TimeData(name="u_ns", shape=model.shape_domain, time_dim=nt,
+                        time_order=time_order, space_order=space_order, save=False,
+                        dtype=model.dtype)
+    u_nosave.pad_time = False
+
+    # Forward wavefield where all timesteps are stored
+    # With checkpointing this should go away <----
+    u_save = TimeData(name="u_s", shape=model.shape_domain, time_dim=nt,
+                      time_order=time_order, space_order=space_order, save=True,
+                      dtype=model.dtype)
+    u_save.pad_time = True
+
```

Comment:
As discussed with @mlange05 and in the meeting, I am not a big fan of all this initializations in the main script. Specially having objects like `Dimensions` that should not be seen by the end user.